lib/query_hook.rb

```diff
@@ -6,10 +6,39 @@ def tempfile_extension
   end
 
   def compile_file_content(r)
-    "#{r.extra}\n#{r.content}\nprint('=> ' + (#{r.query}).inspect)"
+    <<ruby
+    #{r.extra}
+    #{r.content}
+    #{compile_cookie(r.cookie)}
+    #{compile_query(r.query)}
+ruby
   end
 
   def command_line(filename)
     "ruby #{filename}"
   end
-end
\ No newline at end of file
+
+  def compile_query(query)
+    if query.start_with? 'def '
+      "#{query}\nputs \"<function>\""
+    else
+      "print('=> ' + (#{query}).inspect)"
+    end
+  end
+
+  def build_state(cookie)
+    (cookie||[]).map do |statement|
+    <<ruby
```

Comment:
Awesome!

The problem, however, is that this won't properly work when query contains directly or indirectly `puts` statements, since their output will be printed each time you enter a new statement. 

In the wollok runner, I introduced the concept of "MumumukiConsole", which overrides the normal console and replaces by our's, which supress output produced during cookie execution. 

Perhaps we should do something similar here: add code to the query that overrides Kernel's `puts`, so that it acts normally if outside a query, but does nothing otherwise.